src/upload.ts

```diff
@@ -0,0 +1,150 @@
+}
+
+export function flushUploadQueue() {
+  uploadSequentially(uploadQueue);
+}
+
+export function uploadSequentially(uploads: IUploadInfo[]) {
+  if (uploads.length > 0) {
+    let nextUpload = uploads.shift();
+    let uploadNext = (status: number) => {
+      uploadSequentially(uploads);
+    };
+    upload(nextUpload.compressed, nextUpload.raw, uploadNext, uploadNext);
+  }
+}
+
+export function resetUploads() {
+  sentBytesCount = 0;
+  uploadCount = 0;
+  uploadQueue = [];
+  uploadInfos = {};
+  reUploadQueue = [];
+}
+
+function defaultUpload(payload: string, onSuccess?: UploadCallback, onFailure?: UploadCallback) {
+  if (config.uploadUrl.length > 0) {
+    payload = JSON.stringify(payload);
+    let xhr = new XMLHttpRequest();
+    xhr.open("POST", config.uploadUrl);
+    xhr.setRequestHeader("Content-Type", "application/json");
+    xhr.onreadystatechange = () => { onXhrReadyStatusChange(xhr, onSuccess, onFailure); };
+    xhr.send(payload);
+  }
+}
+
+function onXhrReadyStatusChange(xhr: XMLHttpRequest, onSuccess: UploadCallback, onFailure: UploadCallback) {
+  if (xhr.readyState === XMLHttpRequest.DONE) {
+    // HTTP response status documentation:
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
+    if (xhr.status < 200 || xhr.status > 208) {
+      onFailure(xhr.status);
+    } else {
+      onSuccess(xhr.status);
+    }
+  }
+}
+
+function onSuccessDefault(sequenceNumber: number, status: number) {
+  debug(`SUCCESS: Delivered batch #${sequenceNumber}`);
+  let sentData = uploadInfos[sequenceNumber];
+  delete uploadInfos[sequenceNumber];
+  sentBytesCount += sentData.compressed.length;
+
+  if (state === State.Activated && sentBytesCount > config.totalLimit) {
+    let totalByteLimitExceededEventState: ITotalByteLimitExceededEventState = {
+      type: Instrumentation.TotalByteLimitExceeded,
+      bytes: sentBytesCount
+    };
+    instrument(totalByteLimitExceededEventState);
+    teardown();
+    return;
+  }
+
+  reUpload();
+}
+
+function onFailureDefault(sequenceNumber: number, status: number) {
+  debug(`FAILED: Delivery failed for batch #${sequenceNumber} Status: ${status}`);
+  let uploadInfo = uploadInfos[sequenceNumber];
+  uploadInfo.failureCount++;
+  logXhrError(sequenceNumber, status, uploadInfo);
+  if (uploadInfo.failureCount <= config.reUploadLimit) {
+    reUploadQueue.push(sequenceNumber);
+  }
+}
+
+function logXhrError(sequenceNumber: number, status: number, uploadInfo: IUploadInfo) {
+  let events = uploadInfo.raw.events;
+  let xhrErrorState: IXhrErrorEventState = {
+    type: Instrumentation.XhrError,
+    requestStatus: status,
+    sequenceNumber,
+    compressedLength: uploadInfo.compressed.length,
+    rawLength: JSON.stringify(uploadInfo.raw).length,
+    firstEventId: getEventId(events[0]),
+    lastEventId: getEventId(events[events.length - 1]),
+    attemptNumber: uploadInfo.failureCount
+  };
+  instrument(xhrErrorState);
+}
+
+function reUpload() {
```

Comment:
>reUpload [](start = 9, length = 8)

rename to indicate this is for retrying failed uploads

Reply:
How about retryFailedUploads ?

---
In reply to: [168000780](https://github.com/Microsoft/clarity-js/pull/63#discussion_r168000780) [](ancestors = 168000780)